### PR TITLE
[Bug 18646] Use object handles instead of global object pointers

### PR DESCRIPTION
**Pull request by livecodefraser**
Builds on https://github.com/livecode/livecode/pull/4430

### CODE REVIEW

**Comment by montegoulding** [comment]
@livecodefraser this is huge! Are you going to patch the object in `MCSelnode` too I had a crash on that the other day there: http://quality.livecode.com/show_bug.cgi?id=18543

**Comment by montegoulding** [comment]
@livecodefraser regarding GCC's friend warnings do you have any idea what this is about:

```
../../engine/src/card.h:35:18: warning: ‘MCHccard’ is already a friend of ‘MCCard’ [enabled by default]
     friend class MCHccard;
```

File: engine/src/debug.cpp [hunk 1]
```diff
@@ -612,23 +611,25 @@ void MCB_parsewatches(MCExecContext& ctxt, MCStringRef p_input)
...
+                
+                MCObjectHandle t_objecthandle;
+                MCNewAutoNameRef t_handlername;
+                MCNewAutoNameRef t_varname;
+                
+                if (t_resolved_object)
+                    t_objecthandle = t_object.object;
+                
+                /* UNCHECKED */ MCNameCreate(*t_hname, &t_handlername);
+                /* UNCHECKED */ MCNameCreate(*t_vname, &t_varname);
```
Comment (peter-b) [inline engine/src/debug.cpp:628]: Please check these! It should be pretty straightforward because you already have a `t_success` local to hand.

**Review by peter-b — APPROVED** [review]
@livecodefraser This is an impressively large patch!  Hopefully it will make a big improvement to stability (once we get release-mode assertions in).

There's a couple of minor things:
- In a couple of places, you've not fixed `/* UNCHECKED */` calls where you probably could have since you were touching the code anyway.
- Where you've reimplemented list structures to use `MCObjectHandle`, you've added cleaning of the list during method calls that probably wouldn't otherwise be expected to modify the list (such as in `MCSellist::redraw()`).  It might have been better to add a `foreach` abstraction which skips over invalid entries, and use it to implement the non-mutating methods that need to only visit valid members.

I think this is okay to merge without addressing either of the points above.

**Comment by peter-b** [comment]
@livecode-vulcan review ok 6dbdd74

**Comment by livecode-vulcan** [comment]
💙 review by @peter-b ok 6dbdd7445b4203c8ffbc586984059c4df777f018

**Comment by livecode-vulcan** [comment]
😞 test failure 6dbdd7445b4203c8ffbc586984059c4df777f018
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/438)
- try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/443)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/438)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/438)
- try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/444)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/438)
- try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/438)

**Comment by peter-b** [comment]
@livecodefraser As discussed earlier, please rebase this all against the `develop-8.1` branch in order to bring the stability improvements to 8.1.x.  I will file a tracking bug and mark the deletion-pool/object handles bugs as blocked by it.

**Comment by peter-b** [comment]
@livecodefraser I have filed [bug 18646](http://quality.livecode.com/show_bug.cgi?id=18646) to track this.

**Comment by peter-b** [comment]
@livecode-vulcan review ok 23df397

**Comment by livecode-vulcan** [comment]
💙 review by @peter-b ok 23df3970e0216c3ab8546760f1d035e382eef73d

**Comment by peter-b** [comment]
This is now going into `develop-8.1` via #4804

**Pull request closed without merging**